aptos-move/framework/aptos-framework/sources/TokenCoinSwap.move

```diff
@@ -0,0 +1,231 @@
+            };
+            move_to(token_owner, token_store_escrow);
+        }
+    }
+
+    public fun deposite_token_to_escrow(
+        token_owner: &signer,
+        token_id: TokenId,
+        tokens: Token,
+        locked_until_secs: u64
+    ) acquires TokenStoreEscrow {
+        let tokens_in_escrow = &mut borrow_global_mut<TokenStoreEscrow>(
+            Signer::address_of(token_owner)).token_escrows;
+        if (Table::contains(tokens_in_escrow, token_id)) {
+            let dst = &mut Table::borrow_mut(tokens_in_escrow, token_id).token;
+            Token::merge(dst, tokens);
+        } else {
+            let token_escrow = TokenEscrow{
+                token: tokens,
+                locked_until_secs
+            };
+            Table::add(tokens_in_escrow, token_id, token_escrow);
+        };
+    }
+
+    // private function that should be only for internal use
+    fun withdraw_token_from_escrow_internal(
+        token_owner_addr: address,
+        token_id: TokenId,
+        amount: u64
+    ): Token acquires TokenStoreEscrow {
+        let tokens_in_escrow = &mut borrow_global_mut<TokenStoreEscrow>(token_owner_addr).token_escrows;
+        assert!(Table::contains(tokens_in_escrow, token_id), ETOKEN_NOT_IN_ESCROW);
+        let token_escrow = Table::borrow_mut(tokens_in_escrow, token_id);
+        assert!(Timestamp::now_seconds() > token_escrow.locked_until_secs, ETOKEN_CANNOT_MOVE_OUT_OF_ESCROW_BEFORE_LOCKUP_TIME);
+        Token::split(&mut token_escrow.token, amount)
+    }
+
+    public fun withdraw_token_from_escrow(
+        token_owner: &signer,
+        token_id: TokenId,
+        amount: u64
+    ): Token acquires TokenStoreEscrow {
+        withdraw_token_from_escrow_internal(Signer::address_of(token_owner), token_id, amount)
+    }
+
+    public fun cancel_token_listing<CoinType>(
+        token_owner: &signer,
+        token_id: TokenId,
+        token_amount: u64
+    ) acquires TokenListings, TokenStoreEscrow {
+        let listing = &mut borrow_global_mut<TokenListings<CoinType>>(Signer::address_of(token_owner)).listings;
+        // remove the listing entry
+        assert!(Table::contains(listing, token_id), ETOKEN_LISTING_NOT_EXIST);
+        Table::remove(listing, token_id);
+        // get token out of escrow and deposite back to owner token store
+        let tokens = withdraw_token_from_escrow(token_owner, token_id, token_amount);
+        Token::deposit_token(token_owner, tokens);
+    }
+
+    public fun exchange_token_for_coin<CoinType>(
+        token_owner: &signer,
+        token_id: TokenId,
+        token_amout: u64,
+        coin_amount: u64
+    ) {
+        // TODO
+    }
+
+    public fun list_coin_for_swap<CoinType>(
+        coin_owner: &signer,
+        coin_amount: u64,
+        token_id: TokenId,
+        min_token_amount: u64
+    ) {
+        // TODO
+    }
+
+    public fun cancel_coin_listing<CoinType>(coin_owner: &signer, coin_amount: u64) {
+        // TODO
+    }
```

Comment:
I wouldn't worry about this -- just yet.

Reply:
> overall looks good, but I think we should combine forces with the Marketplace that the Martian's were working on: #1204

Yes, will use this basic swap function in https://github.com/aptos-labs/aptos-core/pull/1204.